blockchain/src/chain_store.rs

```diff
@@ -1,21 +1,136 @@
+use std::sync::Arc;
+
 use nimiq_account::RevertInfo;
-use nimiq_block::Block;
+use nimiq_block::{Block, BlockBody, BlockJustification, MacroBlock, MacroBody, MicroBlock};
 use nimiq_blockchain_interface::{BlockchainError, ChainInfo, Direction};
 use nimiq_database::{
     declare_table,
     mdbx::{MdbxDatabase, MdbxReadTransaction, MdbxWriteTransaction, OptionalTransaction},
     traits::{Database, DupReadCursor, ReadCursor, ReadTransaction, WriteCursor, WriteTransaction},
 };
+use nimiq_database_value_derive::DbSerializable;
 use nimiq_hash::Blake2bHash;
 use nimiq_primitives::{policy::Policy, trie::trie_diff::TrieDiff};
+use nimiq_serde::{Deserialize, Serialize};
+use nimiq_transaction::{historic_transaction::HistoricTransactionData, reward::RewardTransaction};
+
+use crate::{history::interface::HistoryInterface, history_store_proxy::HistoryStoreProxy};
 
 declare_table!(HeadTable, "Head", () => Blake2bHash);
 declare_table!(ChainTable, "ChainData", Blake2bHash => ChainInfo);
-declare_table!(BlockTable, "Block", Blake2bHash => Block);
+declare_table!(PushedBlockTable, "PushedBlockTable", Blake2bHash => PushedBlock);
+declare_table!(StoredBlockTable, "StoredBlockTable", Blake2bHash => Block);
 declare_table!(HeightIndex, "HeightIndex", u32 => dup(Blake2bHash));
 declare_table!(RevertTable, "Receipts", u32 => RevertInfo);
 declare_table!(AccountsDiffTable, "AccountsDiff", Blake2bHash => TrieDiff);
 
+/// The non-header content of a block except that transactions are not stored to
+/// optimize blocks storage. This assumes that a block has been pushed and that there
+/// is history associated with this block.
+#[derive(Debug, Serialize, Deserialize, DbSerializable)]
+pub struct PushedBlock {
+    justification: Option<BlockJustification>,
+    #[serde(serialize_with = "PushedBlock::serialize_body")]
+    body: Option<BlockBody>,
+}
```

Comment:
```suggestion
pub enum PushedBlock {
    Macro,
    Micro {
        equivocation_proofs: Vec<EquivocationProof>,
    }
}
```
This could be an alternative way to model this. Technically, we wouldn't need to store anything for macro blocks at all when they are `pushed` (which they should always be).